File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/prolong/ProlongExperimentCommand.java
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration.prolong;

import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;
import pl.allegro.experiments.chi.chiserver.domain.experiments.administration.PermissionsAwareExperimentRepository;

import java.util.Objects;

public class ProlongExperimentCommand {
    private final ExperimentsRepository experimentsRepository;
    private final ProlongExperimentProperties prolongExperimentProperties;
    private final PermissionsAwareExperimentRepository permissionsAwareExperimentRepository;
    private final String experimentId;

    ProlongExperimentCommand(
            ExperimentsRepository experimentsRepository,
            ProlongExperimentProperties prolongExperimentProperties,
            PermissionsAwareExperimentRepository permissionsAwareExperimentRepository,
            String experimentId) {
        Objects.requireNonNull(experimentsRepository);
        Objects.requireNonNull(prolongExperimentProperties);
        Objects.requireNonNull(permissionsAwareExperimentRepository);
        Objects.requireNonNull(experimentId);
        this.experimentId = experimentId;
        this.experimentsRepository = experimentsRepository;
        this.permissionsAwareExperimentRepository = permissionsAwareExperimentRepository;
        this.prolongExperimentProperties = prolongExperimentProperties;
    }

    public void execute() {
        Experiment experiment = permissionsAwareExperimentRepository.getExperimentOrException(experimentId);
        validate(experiment);
        experimentsRepository.save(experiment.prolong(prolongExperimentProperties.getExperimentAdditionalDays()));
    }

    private void validate(Experiment experiment) {
        if (!experiment.isActive()) {
            throw new ProlongExperimentException("Experiment cant be prolonged if it is not ACTIVE");
        }
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/prolong/ProlongExperimentCommand.kt
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration.prolong

import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository
import pl.allegro.experiments.chi.chiserver.domain.experiments.administration.PermissionsAwareExperimentGetter

class ProlongExperimentCommand(
        private val experimentsRepository: ExperimentsRepository,
        private val prolongExperimentProperties: ProlongExperimentProperties,
        private val permissionsAwareExperimentGetter: PermissionsAwareExperimentGetter,
        private val experimentId: String
) {
    fun execute() {
        val experiment = permissionsAwareExperimentGetter.getExperimentOrException(experimentId)
        validate(experiment)
        experimentsRepository.save(experiment.prolong(prolongExperimentProperties.experimentAdditionalDays))
    }

    private fun validate(experiment: Experiment) {
        if (!experiment.isActive()) {
            throw ProlongExperimentException("Experiment is not ACTIVE: ${experimentId}")
        }
        if (prolongExperimentProperties.experimentAdditionalDays <= 0) {
            throw ProlongExperimentException("Experiment additional days must be greater than 0: ${experimentId}")
        }
    }
}